aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -199,11 +237,18 @@ index_select_scale_add(const Tensor &select_indices,
   auto scale_stride = scale.stride(0);
 
   for (int64_t i = 0; i < numel; i++) {
-    auto* src_base = src_data + src_stride0 * select_indices_data[i];
-    auto* output_base = output_data + output_stride0 * add_indices_data[i];
-    auto scale = scale_data[i * scale_stride];
-    for (int64_t j = 0; j < ddim; j++) {
-      output_base[j * output_stride1] += src_base[j * src_stride1] * scale;
+    // We can skip indices equal to padding_idx so they are not included in
+    // the reduction
+    if (select_indices_data[i] != padding_idx) {
+      auto* src_base = src_data + src_stride0 * select_indices_data[i];
+      auto* output_base = output_data + output_stride0 * add_indices_data[i];
+      auto scale = scale_data[i * scale_stride];
+      for (int64_t j = 0; j < ddim; j++) {
+        output_base[j * output_stride1] += src_base[j * src_stride1] * scale;
+      }
+    } else if (bag_size.defined()) {
+      // Decrement bag_size to reflect that the index is padded
+      bag_size_data[add_indices_data[i]]--;
```

Comment:
This branch is only executed if `mode='sum'` and `include_last_offset == True`, so I'll need to add a test that sets those conditions. But I'm having trouble understanding what `include_last_offset` does. The documentation for `EmbeddingBag` says:

```
include_last_offset (bool, optional): if ``True``, :attr:`offsets` has one additional element, where the last element
                              is equivalent to the size of `indices`. This matches the CSR format
```

While the documentation for `embedding_bag` says:

```
include_last_offset (bool, optional): if ``True``, the size of offsets is equal to the number of bags + 1.
                                      The last element is the size of the input, or the ending index
                                      position of the last bag (sequence).
```

I'm having trouble understanding what either of those mean.

So I decided to just try it out to see what it does. I'm using pytorch 1.7.0 for this:

```
>>> torch.manual_seed(0)
<torch._C.Generator object at 0x7fcd9b439410>
>>> a = torch.randn(5, 4)
>>> a
tensor([[-1.1258, -1.1524, -0.2506, -0.4339],
        [ 0.5988, -1.5551, -0.3414,  1.8530],
        [ 0.4681, -0.1577,  1.4437,  0.2660],
        [ 1.3894,  1.5863,  0.9463, -0.8437],
        [ 0.9318,  1.2590,  2.0050,  0.0537]])
>>> torch.nn.functional.embedding_bag(a, torch.tensor([[0, 1], [3, 4], [2, 3]], dtype=torch.long), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]])
>>> torch.nn.functional.embedding_bag(a, torch.tensor([[0, 1], [3, 4], [2, 3]], dtype=torch.long), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899]])
```

With the above example, it looks like `include_last_offset=True` just removes the last row from the result, and no `offsets` input was required. Trying it with CUDA, we get something different:

```
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([[0, 1], [3, 4], [2, 3]], dtype=torch.long).cuda(), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]], device='cuda:0')
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([[0, 1], [3, 4], [2, 3]], dtype=torch.long).cuda(), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 4.1787,  4.2740,  5.3412, -1.3676]], device='cuda:0')
```

This time, `include_last_offset=True` seems to be combining the last two rows of the output.

But I figured, maybe `include_last_offset` is only meant to work properly if `offsets` is given. But I get the exact same behavior as before:

```
>>> torch.nn.functional.embedding_bag(a, torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long), torch.tensor([0, 2, 4]), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]])
>>> torch.nn.functional.embedding_bag(a, torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long), torch.tensor([0, 2, 4]), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899]])
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long).cuda(), torch.tensor([0, 2, 4]).cuda(), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]], device='cuda:0')
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long).cuda(), torch.tensor([0, 2, 4]).cuda(), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 4.1787,  4.2740,  5.3412, -1.3676]], device='cuda:0')
```

Then it occurred to me that the size of the output does follow the documentation when `include_last_offset=True`: `the size of the offsets is equal to the number of bags + 1`. Since I'm using a size `(3, )` offsets input, we get two bags in the .

But the other part of the description(s) for `include_last_offset` is still confusing to me: `The last element is the size of the input, or the ending index position of the last bag` or `the last element is equivalent to the size of indices`. My interpretation of this is: "`offsets[-1]` indicates the number of elements in the `indices` input". So it seems like there should only be one valid value, `offsets[-1] == indices.numel()`. If so, then I should add an extra element, 6, to the end of the `offsets` inputs from the code samples above. When I do that, I get matching results with `include_last_offset=False` and `include_last_offset=True`.

For CPU:
```
>>> torch.nn.functional.embedding_bag(a, torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long), torch.tensor([0, 2, 4]), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]])
>>> torch.nn.functional.embedding_bag(a, torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long), torch.tensor([0, 2, 4, 6]), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]])
```

And CUDA gives the same results:
```
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long).cuda(), torch.tensor([0, 2, 4]).cuda(), mode='sum')
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]], device='cuda:0')
>>> torch.nn.functional.embedding_bag(a.cuda(), torch.tensor([0, 1, 3, 4, 2, 3], dtype=torch.long).cuda(), torch.tensor([0, 2, 4, 6]).cuda(), mode='sum', include_last_offset=True)
tensor([[-0.5270, -2.7075, -0.5919,  1.4191],
        [ 2.3212,  2.8453,  2.9513, -0.7899],
        [ 1.8575,  1.4286,  2.3900, -0.5776]], device='cuda:0')
```

If my interpretation is correct, I still don't understand a few important things:

* Why am I allowed to set `include_last_offset=True` if I don't give an `offsets` input?
* Is the behavior consistent with the confusing name `include_last_offset`?
* Why am I allowed to set `offsets[-1] != indices.numel()`?
* What purpose does this feature serve?

One of the two descriptions of `include_last_offset` says `This matches the CSR format`. Still, I don't see what `include_last_offset` offers if the inputs are sparse. But perhaps it would become clear if I had a better grasp on sparse tensors.

What's your take on this @jbschlosser ?

Reply:
Sorry for the long post, I'm not sure if I could describe my confusion more concisely :stuck_out_tongue_closed_eyes:

Reply:
Ah ok, I found this issue explaining why `include_last_offset` exists: #29019

Its whole purpose is to avoid needing a special case for the last element of `offsets` when `embedding_bag` calculates the size of each bag, when `input` is 1-D. This apparently speeds up parallelism. It seems like `offsets[-1]` should always be equal to `input.size(0)` if `include_last_offset=True`, so I filed #52851 to fix that.

Anyway, I now understand what `include_last_offset` is supposed to do, so I'll add test cases to exercise these branches.

Reply:
Yes, `include_last_offset` is only used in the 1D case, so later on we could also consider adding an error if it's given for the 2D case